File: Advanced Robotics Course/Chapter8/Lesson5/Chapter8_Lesson5.m
function q = grasp_quality_example()
    % Two contact points on a box
    p1 = [0.0;  0.03; 0.0];
    p2 = [0.0; -0.03; 0.0];

    G = buildGraspMap({p1, p2});
    s = svd(G);
    q = min(s);
    fprintf('sigma_min(G) = %f\n', q);
end

function S = skew(v)
    x = v(1); y = v(2); z = v(3);
    S = [ 0, -z,  y;
          z,  0, -x;
         -y,  x,  0];
end

function G = buildGraspMap(contacts)
    m = numel(contacts);
    G_top = zeros(3, 3*m);
    G_bottom = zeros(3, 3*m);
    for i = 1:m
        idx = 3*(i-1) + 1;
        G_top(:, idx:idx+2) = eye(3);
        G_bottom(:, idx:idx+2) = skew(contacts{i});
    end
    G = [G_top; G_bottom];
end
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter8/Lesson5/Chapter8_Lesson5.java
import org.ejml.simple.SimpleMatrix;
import java.util.ArrayList;
import java.util.List;

class Vec3 {
    public double x, y, z;
    public Vec3(double x, double y, double z) {
        this.x = x; this.y = y; this.z = z;
    }
}

public class GraspQuality {

    public static SimpleMatrix skew(Vec3 v) {
        double[][] data = {
                { 0.0,   -v.z,  v.y },
                { v.z,    0.0, -v.x },
                { -v.y,   v.x,  0.0 }
        };
        return new SimpleMatrix(data);
    }

    public static SimpleMatrix buildGraspMap(List<Vec3> contacts) {
        int m = contacts.size();
        SimpleMatrix G_top = new SimpleMatrix(3, 3 * m);
        SimpleMatrix G_bottom = new SimpleMatrix(3, 3 * m);
        G_top.zero(); G_bottom.zero();

        for (int i = 0; i < m; ++i) {
            int idx = 3 * i;
            // Set identity block
            for (int r = 0; r < 3; ++r) {
                G_top.set(r, idx + r, 1.0);
            }
            SimpleMatrix S = skew(contacts.get(i));
            G_bottom.insertIntoThis(0, idx, S);
        }
        // Stack vertically
        SimpleMatrix G = new SimpleMatrix(6, 3 * m);
        G.insertIntoThis(0, 0, G_top);
        G.insertIntoThis(3, 0, G_bottom);
        return G;
    }

    public static double epsilonSurrogate(SimpleMatrix G) {
        SimpleMatrix[] svd = G.svd();
        SimpleMatrix S = svd[1]; // diagonal singular values
        double minSigma = Double.POSITIVE_INFINITY;
        int n = Math.min(S.numRows(), S.numCols());
        for (int i = 0; i < n; ++i) {
            double sigma = S.get(i, i);
            if (sigma < minSigma) {
                minSigma = sigma;
            }
        }
        return minSigma;
    }

    public static void main(String[] args) {
        List<Vec3> contacts = new ArrayList<>();
        contacts.add(new Vec3(0.0,  0.03, 0.0));
        contacts.add(new Vec3(0.0, -0.03, 0.0));

        SimpleMatrix G = buildGraspMap(contacts);
        double q = epsilonSurrogate(G);
        System.out.println("sigma_min(G) = " + q);
    }
}
      
